lib/SILOptimizer/Utils/OwnershipOptUtils.cpp

```diff
@@ -684,65 +737,244 @@ SILBasicBlock::iterator OwnershipRAUWUtility::perform() {
 }
 
 //===----------------------------------------------------------------------===//
-//                          Ownership Fixup Context
+//                     Interior Pointer Operand Rebasing
 //===----------------------------------------------------------------------===//
 
-// All callers of our RAUW routines must ensure that their values return true
-// from this.
-bool OwnershipFixupContext::canFixUpOwnershipForRAUW(SILValue oldValue,
-                                                     SILValue newValue) {
-  auto newOwnershipKind = newValue.getOwnershipKind();
+namespace {
 
-  // If our new kind is ValueOwnershipKind::None, then we are fine. We
-  // trivially support that. This check also ensures that we can always
-  // replace any value with a ValueOwnershipKind::None value.
-  if (newOwnershipKind == OwnershipKind::None)
-    return true;
+/// Clone all projections and casts on the access use-def chain until either the
+/// specified predicate is true or the access base is reached.
+///
+/// This will not clone ref_element_addr or ref_tail_addr because those aren't
+/// part of the access chain.
+class InteriorPointerAddressRebaseUseDefChainCloner
```

Comment:
@atrick I did this based off of https://github.com/apple/swift/blob/a6b85522f5d24a85c2dba89ff17566fb823fb8d1/include/swift/SIL/MemAccessUtils.h#L1495. Did I do this correctly?